sui_programmability/framework/sources/TestScenario.move

```diff
@@ -142,9 +149,33 @@ module Sui::TestScenario {
     /// Same as `remove_object`, but returns the object of type `T` with object ID `id`.
     /// Should only be used in cases where current tx sender has more than one object of
     /// type `T` in their inventory.
-    public fun remove_object_by_id<T: key>(_scenario: &mut Scenario, _id: ID): T {
-        // TODO: implement me
-        abort(100)
+    /// TODO: Perhaps also define can_remove_object_by_id? How to avoid code duplication?
```

Comment:
I think you can at least extract the first part into a helper function `get_object_by_id`

```
public fun remove_object_by_id<T: key>(scenario: &mut Scenario, id: ID): T {
        let object_opt = get_object_by_id<T>(scenario, id);

        assert!(Option::is_some(&object_opt), EOBJECT_ID_NOT_FOUND);
        let object = Option::extract(&mut object_opt);
        Option::destroy_none(object_opt);

        assert!(!Vector::contains(&scenario.removed, &id), EALREADY_REMOVED_OBJECT);
        Vector::push_back(&mut scenario.removed, id);

        object
    }

    /// Return `true` if a call to `remove_object_by_id<T>(scenario, id)` will succeed
    public fun can_remove_object_by_id<T: key>(scenario: &mut Scenario, id: ID): bool {
        if (Vector::contains(&scenario.removed, &id)) {
            return false
        };
        let object_opt = get_object_by_id(scenario, id);
        if (Option::is_some(&object_opt)) {
            let object = Option::extract<T>(&mut object_opt);
            Option::destroy_none(object_opt);
            delete_object_for_testing(object);
            return true
        };
        Option::destroy_none(object_opt);
        false
    }
    
   fun get_object_by_id<T: key>(scenario: &mut Scenario, id: ID): Option<T> {
        let sender = sender(scenario);
        let objects: vector<T> = get_inventory<T>(
            sender,
            last_tx_start_index(scenario)
        );
        let object_opt = Option::none();
        while (!Vector::is_empty(&objects)) {
            let element = Vector::pop_back(&mut objects);
            if (ID::id(&element) == &id) {
                assert!(Option::is_none(&object_opt), EDUPLICATE_OBJCET_ID_FOUND);
                Option::fill(&mut object_opt, element);
            } else {
                delete_object_for_testing(element);
            }
        };
        Vector::destroy_empty(objects);
        object_opt
    }

```

Reply:
One caveat: `get_object_by_id` should not trigger any assertions though. I will think about it more.